Implement Specfile.set_tag()

nforro opened the pull request:
This allows setting value of a tag while attempting to keep the original value intact in case it contains local macro substitutions and updating the corresponding macro definitions instead.

---

A (simple) real-world example:

```spec
Name: postgresql
%global majorversion 14
Version: %{majorversion}.1
```

In this case, `spec.set_tag("Version", "15.1")` would update the `%majorversion` macro, leaving the value of `Version` untouched.

---

I plan to add some integration tests as well.

jpopelka commented on specfile/macro_definitions.py:
```diff
@@ -0,0 +1,209 @@
+            data = super().__getattribute__("data")
+        except AttributeError:
+            return False
+        return name in [md.name for md in data]
```
You don't have to create a whole list before checking, do you?
```suggestion
        return any(md.name == name for md in data)
```

nforro commented:
@FrNecas The PR is now ready for review, I've pretty much rewritten the `ValueParser.construct_regex()` method in order to be able to replace parts of the original value while keeping local macro substitutions intact. I have a bit more changes in mind, but I'll keep that for another PR, including integration tests.

nforro commented:
I've renamed the method to `update_tag()` and added another, more general method `update_value()` for updating any value within a spec file (so it can be used for example to update arguments of `%setup` macro). I've also added `protected_entities` argument that allows to "lock" certain tags and macro definitions to prevent them from being modified.

TomasTomecek reviewed (approved):
Wow, this is wicked!

FrNecas commented on specfile/specfile.py:
```diff
@@ -514,3 +515,118 @@ def add_patch(
+                    if entities[grp]["locked"]:
+                        # avoid infinite recursion
+                        return requested_value
+                    entities[grp]["locked"] = True
```
Small nitpick, the lock-unlock would be suited for a `try-finally` so that we always unlock even if an exception is thrown but it probably does not really matter here (if an exception arises, we are probably not handling it anywhere right now so no point in reseting the lock)

nforro replied:
Right, I decided to leave out `try-finally` in the end, but it's probably more future-proof.

FrNecas reviewed (approved):
Wow, that's quite magical but nice code! Just a few nitpicks, Icouldn't really spot any errors in the logic, I guess we will have to try this out on a few packages to test it thoroughly but I feel like it will be more robust than the algorithm in rebase-helper

nforro commented on specfile/specfile.py:
```diff
@@ -514,3 +515,96 @@ def add_patch(
+        with self.tags() as tags:
+            entities.update(
+                {
+                    t.name: {"value": t.value, "type": type(t), "changed": False}
```
That was my first thought, but then I decided against it because it seemed to me it would lenghten the code for little benefit. But let me try that.

nforro replied:
Ops, dataclasses are not in Python 3.6 :frowning:

nforro replied:
There is a plan to drop Python 3.6/EPEL 8 support, so maybe we should do that right now. What do you think?

FrNecas replied:
Oh, right, I didn't realize that. If we really wanted to use dataclasses, we could do that using a package, in tmt we support epel8 and use dataclasses (https://github.com/teemtee/tmt/blob/main/setup.py#L56)

nforro commented on specfile/macro_definitions.py:
```diff
@@ -0,0 +1,213 @@
+            Constructed instance of `MacroDefinitions` class.
+        """
+        md_regex = re.compile(
+            r"^(\s*)(%(?:global|define))(\s+)(\w+(?:\(.*?\))?)(\s+)(.*?)(\s*|\\)$"
```
You mean verbose, right? I'll do that.

nforro commented on specfile/specfile.py:
```diff
@@ -514,3 +515,96 @@ def add_patch(
+            if m:
+                d = m.groupdict()
+                for k, v in d.items():
+                    if k.startswith("sub_"):
```
I'm not sure if the new name is self-explanatory enough, but if not, I'm open to suggestions :slightly_smiling_face:
